ramips: some GPIO fixes and improvements

DragonBluep opened the pull request.

namiltd commented on target/linux/ramips/mt7621/base-files/etc/board.d/03_gpio_switches:
```diff
@@ -20,25 +20,25 @@ tplink,eap615-wall-v1)
-	ucidef_add_gpio_switch "poe_power_port2" "PoE Power Port2" "402"
-	ucidef_add_gpio_switch "poe_power_port3" "PoE Power Port3" "403"
-	ucidef_add_gpio_switch "poe_power_port4" "PoE Power Port4" "404"
+	ucidef_add_gpio_switch "poe_power_port0" "PoE Power Port0" "608"
```
Why +208?

john-tho commented:
Thank you for fixing these gpio numbers for kernel 6.6.
None of them look critical enough to _need_ a migration script (for sysupgrade from old kernel), so maybe we just make a note to mention this in the next stable release notes.

---

For long term, still wanting to move away from using numbers, but no one has said they are working on anything yet.

Hacky options:
- I think `gpio_switch` could be adjusted to work with named gpios (our `gpio-export` downstream device) fairly simply:
https://github.com/openwrt/openwrt/blob/72cdba7b1a5a573d6b34c8fe462788ee9c951f99/package/base-files/files/etc/init.d/gpio_switch#L25-L26
- It could also be adjusted to grep `/sys/kernel/debug/gpio` for `gpio-line-names` names

- I discovered today [userspace regulator (introduced with kernel 6.2) worked well](https://github.com/openwrt/openwrt/pull/15765#discussion_r1649833286), but we would need to adjust `gpio_switch` (or an alternative to enable or disable regulator from OpenWrt config) to work with that.

Previous discussions:
- It seems [a (libpgiod) simple gpio daemon has been thought about](https://lore.kernel.org/linux-gpio/CAMRc=MedcewoMgykmvB1PHhZNVHMVsPJwzuWhhng3bfV7S7OzA@mail.gmail.com/), but is not near-term.
- Has been mentioned in OpenWrt a few times recently:
  - https://lists.openwrt.org/pipermail/openwrt-devel/2024-March/042448.html
  - https://github.com/openwrt/openwrt/pull/14896#issuecomment-2002162607
  - https://github.com/openwrt/openwrt/pull/15450#issuecomment-2105187026

namiltd commented:
Tested on TP-Link Archer C5 v4 (mt7620, tplink,archer-c5-v4). Works without any problems.

namiltd commented:
I checked this patch for kernel 6.6.36 (https://github.com/openwrt/openwrt/pull/15818) on Archer C5 V4 and surprisingly gpio stopped working:
```
[    0.435577] rt2880_gpio 10000600.gpio: registering 24 gpios
[    0.441013] rt2880_gpio 10000600.gpio: registering 24 irq handlers
[    0.447030] gpio gpiochip0: Static allocation of GPIO base is deprecated, use dynamic allocation.
[    0.456008] rt2880_gpio 10000638.gpio: registering 16 gpios
[    0.461469] rt2880_gpio 10000638.gpio: registering 16 irq handlers
[    0.467427] gpio gpiochip1: Static allocation of GPIO base is deprecated, use dynamic allocation.
[    0.476044] gpio gpiochip1: (10000638.gpio): GPIO integer space overlap, cannot add chip
[    0.483893] gpiochip_add_data_with_key: GPIOs 0..15 (10000638.gpio) failed to register, -16
[    0.492012] rt2880_gpio: probe of 10000638.gpio failed with error -16
[    0.498352] rt2880_gpio 10000660.gpio: registering 32 gpios
[    0.503754] rt2880_gpio 10000660.gpio: registering 32 irq handlers
[    0.509717] gpio gpiochip1: Static allocation of GPIO base is deprecated, use dynamic allocation.
[    0.518337] gpio gpiochip1: (10000660.gpio): GPIO integer space overlap, cannot add chip
[    0.526185] gpiochip_add_data_with_key: GPIOs 0..31 (10000660.gpio) failed to register, -16
[    0.534305] rt2880_gpio: probe of 10000660.gpio failed with error -16
```
Can you check this on your router?

namiltd commented on target/linux/ramips/patches-6.6/802-GPIO-MIPS-ralink-add-gpio-driver-for-ralink-SoC.patch:
```diff
@@ -363,12 +360,6 @@ Cc: linux-gpio@vger.kernel.org
-+	if (gpiobase)
-+		rg->chip.base = be32_to_cpu(*gpiobase);
-+	else
-+		rg->chip.base = -1;
```
Line 370 should remain because after removing it rg->chip.base is not initialized which causes incorrect behavior for the 6.6.36 kernel.
After leaving it, the GPIOs load correctly:
```
[    0.435617] rt2880_gpio 10000600.gpio: registering 24 gpios
[    0.441052] rt2880_gpio 10000600.gpio: registering 24 irq handlers
[    0.447495] rt2880_gpio 10000638.gpio: registering 16 gpios
[    0.452902] rt2880_gpio 10000638.gpio: registering 16 irq handlers
[    0.459245] rt2880_gpio 10000660.gpio: registering 32 gpios
[    0.464705] rt2880_gpio 10000660.gpio: registering 32 irq handlers
[    0.471145] rt2880_gpio 10000688.gpio: registering 1 gpios
[    0.476469] rt2880_gpio 10000688.gpio: registering 1 irq handlers
```

namiltd commented on target/linux/ramips/patches-6.6/802-GPIO-MIPS-ralink-add-gpio-driver-for-ralink-SoC.patch:
```diff
@@ -363,12 +360,6 @@ Cc: linux-gpio@vger.kernel.org
-+	if (gpiobase)
-+		rg->chip.base = be32_to_cpu(*gpiobase);
-+	else
-+		rg->chip.base = -1;
```
Line 370 should remain because after removing it rg->chip.base is not initialized which causes incorrect behavior for the 6.6.36 kernel.
After leaving it, the GPIOs load correctly:
```
[    0.435617] rt2880_gpio 10000600.gpio: registering 24 gpios
[    0.441052] rt2880_gpio 10000600.gpio: registering 24 irq handlers
[    0.447495] rt2880_gpio 10000638.gpio: registering 16 gpios
[    0.452902] rt2880_gpio 10000638.gpio: registering 16 irq handlers
[    0.459245] rt2880_gpio 10000660.gpio: registering 32 gpios
[    0.464705] rt2880_gpio 10000660.gpio: registering 32 irq handlers
[    0.471145] rt2880_gpio 10000688.gpio: registering 1 gpios
[    0.476469] rt2880_gpio 10000688.gpio: registering 1 irq handlers
```

DragonBluep replied:
Fixed.

DragonBluep committed:
ramips: gpio: always use dynamic GPIO numberspace base

The new numberspace base starts from 512 instead of 0. The number
base seems come from the kernel symbol CONFIG_GPIOLIB_FASTPATH_LIMIT.

Suppress warning:

gpio gpiochip0: Static allocation of GPIO base is deprecated, use dynamic allocation.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep committed:
ramips: gpio: remove header "gpio.h"

There are no source files using it.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep committed:
ramips: mt7621: fix GPIO switch pin number

The numberspace base has been changed since 6.6 kernel:

chip_num   chip0   chip1   chip2 (32 gpios per bank)
old base    480     448     416
new base    512     544     576

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep committed:
ramips: gpio: convert to the generic GPIO driver

Drop most of the code in favor of the generic MMIO GPIO driver.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep committed:
ramips: gpio: use devm_platform_ioremap_resource()

Simplify the code.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

namiltd commented:
Successfully tested on Archer C5 v4 on kernel 6.6.36.
